SftpLibsshAdapter/SFTPConnection.cpp

```diff
@@ -0,0 +1,269 @@
+		const mode_t  DEFAULT_FILE_PERMISSION = 0; 
+
+		std::string readFileContent(const std::string filename)
+		{
+			std::string content = "";
+			try
+			{
+				std::ifstream file(filename);
+
+				file.seekg(0, std::ios::end);
+				content.reserve((unsigned int)file.tellg());
+				file.seekg(0, std::ios::beg);
+
+				content.assign((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+			}
+			catch (...) {}
+			return content;
+		}
+
+	}
+
+	SFTPConnection::~SFTPConnection()
+	{
+		close();
+	}
+
+	void SFTPConnection::connect(const std::string& ip,
+									unsigned int port,
+									const std::string& username,
+									const std::string& pubKeyFile,
+									const std::string& privKeyFile,
+									const std::function<std::string()>& getPrivKeyPassPhraseFn,
+									const std::vector<std::string>& serverFingerPrints)
+	{
+		try
+		{
+			sshConnect(ip, port);
+			verifySFTPServerIdentity(serverFingerPrints);
+			sshAuthorize(username, pubKeyFile, privKeyFile, getPrivKeyPassPhraseFn());
+			sftpConnect();
+		}
+		catch (ISFTPClientAdapter::Exception exc)
+		{
+			close();
+			throw exc;
+		}
+	}
+
+	void SFTPConnection::sshConnect(const std::string& ip, unsigned int port)
+	{
+		m_sshSession = ssh_new();
+		if (!m_sshSession)
+		{
+			throw SSHSessionCannotCreateException(ssh_get_error(m_sshSession));
+		}
+
+		ssh_options_set(m_sshSession, SSH_OPTIONS_HOST, ip.c_str());
+		ssh_options_set(m_sshSession, SSH_OPTIONS_PORT, &port);
+
+		if (ssh_connect(m_sshSession) != SSH_OK)
+		{
+			throw ConectionFailedException(ssh_get_error(m_sshSession));
+		}
+	}
+
+	void SFTPConnection::sshAuthorize(const std::string& username, const std::string& pubKeyFile,
+									  const std::string& privKeyFile, const std::string& privKeyPassPhrase)
+	{
+		ssh_key publickey = nullptr;
+		ssh_key privatekey = nullptr;
+		std::exception_ptr raisedException = nullptr;
+
+		try
+		{
+			// IMPORTANT only base64 format public key are accepted as file content
+			std::string publicKeyFileContent = readFileContent(pubKeyFile);
+			if (ssh_pki_import_pubkey_base64(publicKeyFileContent.c_str(), SSH_KEYTYPE_RSA, &publickey) != SSH_OK)
+			{
+				throw InvalidPublicKeyException(ssh_get_error(m_sshSession));
+			}
+
+			if (ssh_userauth_try_publickey(m_sshSession, username.c_str(), publickey) != SSH_OK)
+			{
+				throw AuthenticationFailedException(ssh_get_error(m_sshSession));
+			}
+
+			// IMPORTANT only openSSH format private key file are accepted
+			if (ssh_pki_import_privkey_file(privKeyFile.c_str(), privKeyPassPhrase.c_str(), NULL, NULL, &privatekey) != SSH_AUTH_SUCCESS)
+			{
+				throw InvalidPrivateKeyException(ssh_get_error(m_sshSession));
+			}
+
+			if (ssh_userauth_publickey(m_sshSession, username.c_str(), privatekey) != SSH_OK)
+			{
+				throw AuthenticationFailedPrivateException(ssh_get_error(m_sshSession));
+			}
+		}
+		catch (ISFTPClientAdapter::Exception exc)
```

Comment:
I think the throw with try-catching with rethrow to be able to free resources is a bit confusing.
Why don't you do a function like `freeSSHKeys(publickey, privatekey)` and call it before every throw and at the end of the function (that you reach if nothing threw).

This way you avoid the catch